openc3-cosmos-init/plugins/packages/openc3-cosmos-tool-cmdtlmserver/src/tools/CmdTlmServer/TargetsTab.vue

```diff
@@ -63,25 +149,149 @@ export default {
+          this.commandAuthority = true
+        }
+      })
+      // Get the initial command authority settings
+      Api.get('/openc3-api/cmdauth').then((response) => {
+        this.cmdAuth = response.data
+      })
+      // Create a cable to the SystemEventsChannel so we can maintain
+      // state with the backend
+      this.cable
+        .createSubscription('SystemEventsChannel', window.openc3Scope, {
+          received: (data) => {
+            this.cable.recordPing()
+            this.handleMessages(data)
+          },
+        })
+        .then((systemSubscription) => {
+          this.systemSubscription = systemSubscription
+        })
     }
   },
+  destroyed() {
+    if (this.systemSubscription) {
+      this.systemSubscription.unsubscribe()
+    }
+    this.cable.disconnect()
+  },
   methods: {
+    // Enterprise only
+    handleMessages(data) {
+      data.forEach((message) => {
+        let event = JSON.parse(message['event'])
+        switch (event.type) {
+          case 'scope':
+            if (window.openc3Scope === event.name) {
+              this.commandAuthority = event.command_authority
+            }
+            break
+          case 'cmd_auth_take':
+            this.cmdAuth[event.name] = event
+            break
+          case 'cmd_auth_release':
+            delete this.cmdAuth[event.name]
+            break
+          // There is also 'role' but we don't care
+        }
+      })
+    },
+    tooltipHandler(method) {
+      if (this.enterprise) {
+        if (this.commandAuthority) {
+          this[method]()
+        } else {
+          window.open('/tools/admin/scopes', '_blank')
+        }
+      } else {
+        this.showUpgradeToEnterpriseDialog = true
+      }
+    },
     update() {
       if (this.tabId != this.curTab) return
-      this.api.get_target_interfaces().then((info) => {
-        this.data = []
-        for (let x of info) {
-          this.data.push({
-            name: x[0],
-            interface: x[1],
-          })
+      this.api.get_target_interfaces().then(async (info) => {
+        for (let i = 0; i < info.length; i++) {
+          this.data[i].name = info[i][0]
+          this.data[i].interface = info[i][1]
+          if (this.enterprise) {
+            if (this.cmdAuth[this.data[i].name]) {
+              this.data[i].username =
+                this.cmdAuth[this.data[i].name]['username']
+            } else {
+              this.data[i].username = ''
+            }
+          }
         }
       })
     },
+    take(target_name) {
+      Api.post('/openc3-api/cmdauth/take', {
+        data: {
+          target_name: target_name,
+        },
+      })
+    },
+    release(target_name) {
+      Api.post('/openc3-api/cmdauth/release', {
+        data: {
+          target_name: target_name,
+        },
+      })
+    },
+    takeAll() {
+      Api.post('/openc3-api/cmdauth/take-all', {
+        data: {
+          target_names: this.data.map((target) => target.name),
```

Comment:
Why does this need a list of target names?

Reply:
Probably doesn't but I already had the list in the frontend. Probably more robust to ask for the list in the backend.